lib/services/local/access_graph.go

```diff
@@ -0,0 +1,178 @@
+func (k *AccessGraphSecretsService) ListAllAuthorizedKeys(ctx context.Context, pageSize int, pageToken string) ([]*accessgraphsecretspb.AuthorizedKey, string, error) {
+	out, next, err := k.authorizedKeysSvc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// ListAuthorizedKeysForServer lists all authorized keys for a given hostID.
+func (k *AccessGraphSecretsService) ListAuthorizedKeysForServer(ctx context.Context, hostID string, pageSize int, pageToken string) ([]*accessgraphsecretspb.AuthorizedKey, string, error) {
+	if hostID == "" {
+		return nil, "", trace.BadParameter("server name is required")
+	}
+	svc := k.authorizedKeysSvc.WithPrefix(hostID)
+	out, next, err := svc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// UpsertAuthorizedKey upserts a new authorized key.
+func (k *AccessGraphSecretsService) UpsertAuthorizedKey(ctx context.Context, in *accessgraphsecretspb.AuthorizedKey) (*accessgraphsecretspb.AuthorizedKey, error) {
+	svc := k.authorizedKeysSvc.WithPrefix(in.Spec.HostId)
+	out, err := svc.UpsertResource(ctx, in)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return out, nil
+}
+
+// DeleteAuthorizedKey deletes a specific authorized key.
+func (k *AccessGraphSecretsService) DeleteAuthorizedKey(ctx context.Context, hostID, name string) error {
+	svc := k.authorizedKeysSvc.WithPrefix(hostID)
+	return trace.Wrap(svc.DeleteResource(ctx, name))
+}
+
+// DeleteAllAuthorizedKeys deletes all authorized keys.
+func (k *AccessGraphSecretsService) DeleteAllAuthorizedKeys(ctx context.Context) error {
+	return trace.Wrap(k.authorizedKeysSvc.DeleteAllResources(ctx))
+}
+
+// ListAllPrivateKeys lists all private keys stored in the backend.
+func (k *AccessGraphSecretsService) ListAllPrivateKeys(ctx context.Context, pageSize int, pageToken string) ([]*accessgraphsecretspb.PrivateKey, string, error) {
+	out, next, err := k.privateKeysSvc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// ListPrivateKeysForDevice lists all private keys for a given deviceID.
+func (k *AccessGraphSecretsService) ListPrivateKeysForDevice(ctx context.Context, deviceID string, pageSize int, pageToken string) ([]*accessgraphsecretspb.PrivateKey, string, error) {
+	if deviceID == "" {
+		return nil, "", trace.BadParameter("server name is required")
+	}
+	svc := k.privateKeysSvc.WithPrefix(deviceID)
+	out, next, err := svc.ListResources(ctx, pageSize, pageToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return out, next, nil
+}
+
+// UpsertPrivateKey upserts a new private key.
+func (k *AccessGraphSecretsService) UpsertPrivateKey(ctx context.Context, in *accessgraphsecretspb.PrivateKey) (*accessgraphsecretspb.PrivateKey, error) {
+	svc := k.privateKeysSvc.WithPrefix(in.Spec.DeviceId)
```

Comment:
Wonder why the deviceId prefix is needed the name consist with user + deviceID and fingerprint hash. 

Does the backend will also have ability to list all keys from particular machine. If not the DeviceID prefix can be removed from backend item path.

Reply:
Since each device reports all the new findings, we need to list the last reported data for a certain device to reconcile. everything happens server side!